File: concepts/02-scalability/python/sharding.py
"""
Sharding Strategies Implementation
Demonstrates different sharding approaches
"""
from typing import Dict, Any, Optional
import hashlib


class RangeBasedSharding:
    """Range-based sharding implementation"""
    
    def __init__(self, num_shards: int):
        self.num_shards = num_shards
        self.shards: Dict[int, Dict] = {i: {} for i in range(num_shards)}
    
    def get_shard(self, key: int) -> int:
        """Get shard for key based on range"""
        return key % self.num_shards
    
    def set(self, key: int, value: Any) -> Dict[str, Any]:
        """Set value"""
        shard_id = self.get_shard(key)
        self.shards[shard_id][key] = value
        return {'shard_id': shard_id, 'key': key, 'value': value}
    
    def get(self, key: int) -> Optional[Any]:
        """Get value"""
        shard_id = self.get_shard(key)
        return self.shards[shard_id].get(key)
    
    def get_stats(self) -> Dict[int, Dict]:
        """Get shard statistics"""
        return {
            i: {
                'size': len(self.shards[i]),
                'keys': list(self.shards[i].keys())
            }
            for i in range(self.num_shards)
        }


class HashBasedSharding:
    """Hash-based sharding implementation"""
    
    def __init__(self, num_shards: int):
        self.num_shards = num_shards
        self.shards: Dict[int, Dict] = {i: {} for i in range(num_shards)}
    
    def hash(self, key: str) -> int:
        """Hash function"""
        return int(hashlib.md5(key.encode()).hexdigest(), 16)
    
    def get_shard(self, key: str) -> int:
        """Get shard for key based on hash"""
        hash_value = self.hash(key)
        return hash_value % self.num_shards
    
    def set(self, key: str, value: Any) -> Dict[str, Any]:
        """Set value"""
        shard_id = self.get_shard(key)
        self.shards[shard_id][key] = value
        return {'shard_id': shard_id, 'key': key, 'value': value}
    
    def get(self, key: str) -> Optional[Any]:
        """Get value"""
        shard_id = self.get_shard(key)
        return self.shards[shard_id].get(key)
    
    def get_stats(self) -> Dict[int, Dict]:
        """Get shard statistics"""
        return {
            i: {'size': len(self.shards[i])}
            for i in range(self.num_shards)
        }


class DirectoryBasedSharding:
    """Directory-based sharding implementation"""
    
    def __init__(self, num_shards: int):
        self.num_shards = num_shards
        self.shards: Dict[int, Dict] = {i: {} for i in range(num_shards)}
        self.directory: Dict[str, int] = {}  # Key -> Shard mapping
    
    def get_shard(self, key: str) -> int:
        """Get shard for key (with directory lookup)"""
        if key in self.directory:
            return self.directory[key]
        
        # Assign to shard (could be based on load, etc.)
        import random
        shard_id = random.randint(0, self.num_shards - 1)
        self.directory[key] = shard_id
        return shard_id
    
    def set(self, key: str, value: Any) -> Dict[str, Any]:
        """Set value"""
        shard_id = self.get_shard(key)
        self.shards[shard_id][key] = value
        return {'shard_id': shard_id, 'key': key, 'value': value}
    
    def get(self, key: str) -> Optional[Any]:
        """Get value"""
        shard_id = self.directory.get(key)
        if shard_id is None:
            return None
        return self.shards[shard_id].get(key)
    
    def rebalance(self, key: str, new_shard_id: int) -> Optional[Dict[str, Any]]:
        """Rebalance: Move key to different shard"""
        old_shard_id = self.directory.get(key)
        if old_shard_id is not None:
            value = self.shards[old_shard_id].pop(key)
            self.shards[new_shard_id][key] = value
            self.directory[key] = new_shard_id
            return {'key': key, 'old_shard_id': old_shard_id, 'new_shard_id': new_shard_id}
        return None
    
    def get_stats(self) -> Dict[int, Dict]:
        """Get shard statistics"""
        return {
            i: {'size': len(self.shards[i])}
            for i in range(self.num_shards)
        }


class ConsistentHashing:
    """Consistent hashing implementation"""
    
    def __init__(self, num_shards: int, virtual_nodes: int = 3):
        self.num_shards = num_shards
        self.virtual_nodes = virtual_nodes
        self.shards: Dict[int, Dict] = {i: {} for i in range(num_shards)}
        self.ring: Dict[int, int] = {}  # Hash value -> Shard ID
        self.sorted_keys: list = []
        
        self._build_ring()
    
    def hash(self, key: str) -> int:
        """Hash function"""
        return int(hashlib.md5(key.encode()).hexdigest(), 16)
    
    def _build_ring(self):
        """Build hash ring"""
        self.ring.clear()
        self.sorted_keys = []
        
        for shard_id in range(self.num_shards):
            for v in range(self.virtual_nodes):
                hash_value = self.hash(f'{shard_id}-{v}')
                self.ring[hash_value] = shard_id
                self.sorted_keys.append(hash_value)
        
        self.sorted_keys.sort()
    
    def get_shard(self, key: str) -> int:
        """Get shard for key using consistent hashing"""
        key_hash = self.hash(key)
        
        # Find first node with hash >= key_hash
        for ring_hash in self.sorted_keys:
            if ring_hash >= key_hash:
                return self.ring[ring_hash]
        
        # Wrap around to first node
        return self.ring[self.sorted_keys[0]]
    
    def set(self, key: str, value: Any) -> Dict[str, Any]:
        """Set value"""
        shard_id = self.get_shard(key)
        self.shards[shard_id][key] = value
        return {'shard_id': shard_id, 'key': key, 'value': value}
    
    def get(self, key: str) -> Optional[Any]:
        """Get value"""
        shard_id = self.get_shard(key)
        return self.shards[shard_id].get(key)
    
    def add_shard(self) -> int:
        """Add new shard"""
        new_shard_id = self.num_shards
        self.num_shards += 1
        self.shards[new_shard_id] = {}
        self._build_ring()  # Rebuild ring
        return new_shard_id
    
    def remove_shard(self, shard_id: int):
        """Remove shard"""
        if shard_id in self.shards:
            # Move data to other shards (simplified)
            data = self.shards[shard_id].copy()
            for key, value in data.items():
                new_shard_id = self.get_shard(key)
                if new_shard_id != shard_id:
                    self.shards[new_shard_id][key] = value
            
            del self.shards[shard_id]
            self._build_ring()  # Rebuild ring
            self.num_shards -= 1
    
    def get_stats(self) -> Dict[int, Dict]:
        """Get shard statistics"""
        return {
            i: {'size': len(self.shards[i])}
            for i in range(self.num_shards)
        }


def demonstrate_sharding():
    """Demonstrate different sharding strategies"""
    print('=== Range-Based Sharding ===\n')
    range_sharding = RangeBasedSharding(3)
    for i in range(1, 11):
        range_sharding.set(i, f'value{i}')
    print('Stats:', range_sharding.get_stats())
    print('Get key 5:', range_sharding.get(5))
    
    print('\n=== Hash-Based Sharding ===\n')
    hash_sharding = HashBasedSharding(3)
    for i in range(1, 11):
        hash_sharding.set(f'user{i}', f'value{i}')
    print('Stats:', hash_sharding.get_stats())
    print('Get user5:', hash_sharding.get('user5'))
    
    print('\n=== Directory-Based Sharding ===\n')
    dir_sharding = DirectoryBasedSharding(3)
    for i in range(1, 11):
        dir_sharding.set(f'key{i}', f'value{i}')
    print('Stats:', dir_sharding.get_stats())
    dir_sharding.rebalance('key5', 2)
    print('After rebalance:', dir_sharding.get_stats())
    
    print('\n=== Consistent Hashing ===\n')
    consistent_hashing = ConsistentHashing(3, 3)
    for i in range(1, 11):
        consistent_hashing.set(f'key{i}', f'value{i}')
    print('Stats:', consistent_hashing.get_stats())
    print('Adding new shard...')
    consistent_hashing.add_shard()
    print('Stats after adding shard:', consistent_hashing.get_stats())


if __name__ == '__main__':
    demonstrate_sharding()

File: concepts/02-scalability/node/sharding.js
/**
 * Sharding Strategies Implementation
 * Demonstrates different sharding approaches
 */

/**
 * Range-Based Sharding
 */
class RangeBasedSharding {
  constructor(numShards) {
    this.numShards = numShards;
    this.shards = new Map();
    this.ranges = [];
    
    // Initialize shards
    for (let i = 0; i < numShards; i++) {
      this.shards.set(i, new Map());
    }
  }

  /**
   * Get shard for key based on range
   */
  getShard(key) {
    // Simple range-based: key % numShards
    const shardId = key % this.numShards;
    return shardId;
  }

  /**
   * Set value
   */
  set(key, value) {
    const shardId = this.getShard(key);
    this.shards.get(shardId).set(key, value);
    return { shardId, key, value };
  }

  /**
   * Get value
   */
  get(key) {
    const shardId = this.getShard(key);
    return this.shards.get(shardId).get(key);
  }

  /**
   * Get shard statistics
   */
  getStats() {
    const stats = {};
    for (let i = 0; i < this.numShards; i++) {
      stats[i] = {
        size: this.shards.get(i).size,
        keys: Array.from(this.shards.get(i).keys())
      };
    }
    return stats;
  }
}

/**
 * Hash-Based Sharding
 */
class HashBasedSharding {
  constructor(numShards) {
    this.numShards = numShards;
    this.shards = new Map();
    
    for (let i = 0; i < numShards; i++) {
      this.shards.set(i, new Map());
    }
  }

  /**
   * Simple hash function
   */
  hash(key) {
    let hash = 0;
    const str = String(key);
    for (let i = 0; i < str.length; i++) {
      const char = str.charCodeAt(i);
      hash = ((hash << 5) - hash) + char;
      hash = hash & hash; // Convert to 32-bit integer
    }
    return Math.abs(hash);
  }

  /**
   * Get shard for key based on hash
   */
  getShard(key) {
    const hashValue = this.hash(key);
    return hashValue % this.numShards;
  }

  /**
   * Set value
   */
  set(key, value) {
    const shardId = this.getShard(key);
    this.shards.get(shardId).set(key, value);
    return { shardId, key, value };
  }

  /**
   * Get value
   */
  get(key) {
    const shardId = this.getShard(key);
    return this.shards.get(shardId).get(key);
  }

  /**
   * Get shard statistics
   */
  getStats() {
    const stats = {};
    for (let i = 0; i < this.numShards; i++) {
      stats[i] = {
        size: this.shards.get(i).size
      };
    }
    return stats;
  }
}

/**
 * Directory-Based Sharding
 */
class DirectoryBasedSharding {
  constructor(numShards) {
    this.numShards = numShards;
    this.shards = new Map();
    this.directory = new Map(); // Key -> Shard mapping
    
    for (let i = 0; i < numShards; i++) {
      this.shards.set(i, new Map());
    }
  }

  /**
   * Get shard for key (with directory lookup)
   */
  getShard(key) {
    if (this.directory.has(key)) {
      return this.directory.get(key);
    }
    
    // Assign to shard (could be based on load, etc.)
    const shardId = Math.floor(Math.random() * this.numShards);
    this.directory.set(key, shardId);
    return shardId;
  }

  /**
   * Set value
   */
  set(key, value) {
    const shardId = this.getShard(key);
    this.shards.get(shardId).set(key, value);
    return { shardId, key, value };
  }

  /**
   * Get value
   */
  get(key) {
    const shardId = this.directory.get(key);
    if (shardId === undefined) {
      return undefined;
    }
    return this.shards.get(shardId).get(key);
  }

  /**
   * Rebalance: Move key to different shard
   */
  rebalance(key, newShardId) {
    const oldShardId = this.directory.get(key);
    if (oldShardId !== undefined) {
      const value = this.shards.get(oldShardId).get(key);
      this.shards.get(oldShardId).delete(key);
      this.shards.get(newShardId).set(key, value);
      this.directory.set(key, newShardId);
      return { key, oldShardId, newShardId };
    }
    return null;
  }

  /**
   * Get shard statistics
   */
  getStats() {
    const stats = {};
    for (let i = 0; i < this.numShards; i++) {
      stats[i] = {
        size: this.shards.get(i).size
      };
    }
    return stats;
  }
}

/**
 * Consistent Hashing
 */
class ConsistentHashing {
  constructor(numShards, virtualNodes = 3) {
    this.numShards = numShards;
    this.virtualNodes = virtualNodes;
    this.shards = new Map();
    this.ring = new Map(); // Hash value -> Shard ID
    this.sortedKeys = [];
    
    // Initialize shards
    for (let i = 0; i < numShards; i++) {
      this.shards.set(i, new Map());
    }
    
    // Create hash ring with virtual nodes
    this.buildRing();
  }

  /**
   * Hash function
   */
  hash(key) {
    let hash = 0;
    const str = String(key);
    for (let i = 0; i < str.length; i++) {
      const char = str.charCodeAt(i);
      hash = ((hash << 5) - hash) + char;
      hash = hash & hash;
    }
    return Math.abs(hash);
  }

  /**
   * Build hash ring
   */
  buildRing() {
    this.ring.clear();
    this.sortedKeys = [];
    
    for (let shardId = 0; shardId < this.numShards; shardId++) {
      for (let v = 0; v < this.virtualNodes; v++) {
        const hashValue = this.hash(`${shardId}-${v}`);
        this.ring.set(hashValue, shardId);
        this.sortedKeys.push(hashValue);
      }
    }
    
    this.sortedKeys.sort((a, b) => a - b);
  }

  /**
   * Get shard for key using consistent hashing
   */
  getShard(key) {
    const keyHash = this.hash(key);
    
    // Find first node with hash >= keyHash
    for (const ringHash of this.sortedKeys) {
      if (ringHash >= keyHash) {
        return this.ring.get(ringHash);
      }
    }
    
    // Wrap around to first node
    return this.ring.get(this.sortedKeys[0]);
  }

  /**
   * Set value
   */
  set(key, value) {
    const shardId = this.getShard(key);
    this.shards.get(shardId).set(key, value);
    return { shardId, key, value };
  }

  /**
   * Get value
   */
  get(key) {
    const shardId = this.getShard(key);
    return this.shards.get(shardId).get(key);
  }

  /**
   * Add new shard
   */
  addShard() {
    const newShardId = this.numShards;
    this.numShards++;
    this.shards.set(newShardId, new Map());
    this.buildRing(); // Rebuild ring
    return newShardId;
  }

  /**
   * Remove shard
   */
  removeShard(shardId) {
    if (this.shards.has(shardId)) {
      // Move data to other shards (simplified)
      const data = this.shards.get(shardId);
      for (const [key, value] of data) {
        const newShardId = this.getShard(key);
        if (newShardId !== shardId) {
          this.shards.get(newShardId).set(key, value);
        }
      }
      
      this.shards.delete(shardId);
      this.buildRing(); // Rebuild ring
      this.numShards--;
    }
  }

  /**
   * Get shard statistics
   */
  getStats() {
    const stats = {};
    for (let i = 0; i < this.numShards; i++) {
      stats[i] = {
        size: this.shards.get(i).size
      };
    }
    return stats;
  }
}

// Example usage
function demonstrateSharding() {
  console.log('=== Range-Based Sharding ===\n');
  const rangeSharding = new RangeBasedSharding(3);
  for (let i = 1; i <= 10; i++) {
    rangeSharding.set(i, `value${i}`);
  }
  console.log('Stats:', rangeSharding.getStats());
  console.log('Get key 5:', rangeSharding.get(5));

  console.log('\n=== Hash-Based Sharding ===\n');
  const hashSharding = new HashBasedSharding(3);
  for (let i = 1; i <= 10; i++) {
    hashSharding.set(`user${i}`, `value${i}`);
  }
  console.log('Stats:', hashSharding.getStats());
  console.log('Get user5:', hashSharding.get('user5'));

  console.log('\n=== Directory-Based Sharding ===\n');
  const dirSharding = new DirectoryBasedSharding(3);
  for (let i = 1; i <= 10; i++) {
    dirSharding.set(`key${i}`, `value${i}`);
  }
  console.log('Stats:', dirSharding.getStats());
  dirSharding.rebalance('key5', 2);
  console.log('After rebalance:', dirSharding.getStats());

  console.log('\n=== Consistent Hashing ===\n');
  const consistentHashing = new ConsistentHashing(3, 3);
  for (let i = 1; i <= 10; i++) {
    consistentHashing.set(`key${i}`, `value${i}`);
  }
  console.log('Stats:', consistentHashing.getStats());
  console.log('Adding new shard...');
  consistentHashing.addShard();
  console.log('Stats after adding shard:', consistentHashing.getStats());
}

if (require.main === module) {
  demonstrateSharding();
}

module.exports = {
  RangeBasedSharding,
  HashBasedSharding,
  DirectoryBasedSharding,
  ConsistentHashing
};

